src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_argent_challenge.cpp

```diff
@@ -104,321 +150,710 @@ enum Spells
+                {
+                    instance->SetBossState(DATA_ARGENT_CHALLENGE, SPECIAL);
+                    events.Reset();
+                    Talk(SAY_DEFEATED_E);
+                }
             }
         }
 
-        void MovementInform(uint32 MovementType, uint32 /*Data*/) override
+        void JustEngagedWith(Unit* who) override
         {
-            if (MovementType != POINT_MOTION_TYPE)
+            if (instance->GetBossState(DATA_ARGENT_CHALLENGE) == SPECIAL)
                 return;
 
-            instance->SetData(BOSS_ARGENT_CHALLENGE_E, DONE);
+            DoCastAOE(SPELL_VENGEANCE);
+            events.ScheduleEvent(EVENT_RADIANCE, urand(7 * IN_MILLISECONDS, 15 * IN_MILLISECONDS));
+            Talk(SAY_AGGRO_E, who);
+            _JustEngagedWith();
+        }
 
-            me->DisappearAndDie();
+        void KilledUnit(Unit* who) override
+        {
+            if (who->GetTypeId() == TYPEID_PLAYER)
+                Talk(SAY_KILL_UNIT_E, who);
         }
 
-        void UpdateAI(uint32 uiDiff) override
+        void UpdateAI(uint32 diff) override
         {
-            if (bDone && uiResetTimer <= uiDiff)
-            {
-                me->GetMotionMaster()->MovePoint(0, 746.87f, 665.87f, 411.75f);
-                bDone = false;
-            } else uiResetTimer -= uiDiff;
+            events.Update(diff);
 
-            if (!UpdateVictim())
+            if (me->HasUnitState(UNIT_STATE_CASTING))
                 return;
 
-            if (uiHammerJusticeTimer <= uiDiff)
+            while (uint32 eventId = events.ExecuteEvent())
             {
-                me->InterruptNonMeleeSpells(true);
-
-                if (Unit* target = SelectTarget(SELECT_TARGET_RANDOM, 0, 250, true))
+                switch (eventId)
                 {
-                    if (target->IsAlive())
+                case EVENT_RADIANCE:
+                    DoCastAOE(SPELL_RADIANCE);
+                    Talk(EMOTE_RADIANCE);
+                    events.ScheduleEvent(EVENT_HAMMER_OF_JUSTICE, urand(20 * IN_MILLISECONDS, 30 * IN_MILLISECONDS));
+                    break;
+                case EVENT_HAMMER_OF_JUSTICE:
+                    if (Unit* target = SelectTarget(SELECT_TARGET_RANDOM, 0, 100.0f, true))
                     {
-                        DoCast(target, SPELL_HAMMER_JUSTICE);
+                        DoCast(target, SPELL_HAMMER_JUSTICE, true);
                         DoCast(target, SPELL_HAMMER_RIGHTEOUS);
+                        Talk(EMOTE_HAMMER_RIGHTEOUS, target);
+                        Talk(SAY_HAMMER_RIGHTEOUS);
                     }
+                    events.ScheduleEvent(EVENT_RADIANCE, urand(7 * IN_MILLISECONDS, 15 * IN_MILLISECONDS));
+                    break;
+                case EVENT_CHALLENGE_DONE:
+                    me->RemoveAllAuras();
+                    me->GetThreatManager().ClearAllThreat();
+                    me->SetFullHealth();
```

Comment:
Should there be an immunity applied here somewhere? Is it intended that players can still hit him?